src/hotspot/share/gc/g1/heapRegionRemSet.cpp

```diff
@@ -124,7 +120,10 @@ void HeapRegionRemSet::add_code_root_locked(nmethod* nm) {
           "not safely locked. CodeCache_lock->owned_by_self(): %s, is_at_safepoint(): %s, _m.owned_by_self(): %s, Thread::current()->is_VM_thread(): %s",
           BOOL_TO_STR(CodeCache_lock->owned_by_self()), BOOL_TO_STR(SafepointSynchronize::is_at_safepoint()),
           BOOL_TO_STR(_m.owned_by_self()), BOOL_TO_STR(Thread::current()->is_VM_thread()));
-  _code_roots.add(nm);
+
+  if (!_code_roots.contains(nm)) { // with this test, we can assert that we do not modify the hash table while iterating over it
```

Comment:
But you have the assert in 'add' too, so put_if_absent is more efficient than calling contains first.

Reply:
When we have an evacuation failure, the code will add an nmethod that is already present. If I guard in `add_code_root_locked`, evacuation failures will not reach the assert in `G1CodeRootSet::add`. Another way to do it is to to do the `contains` within the assert and then somewhat optimising the release build. Would you prefer that?